File: src/lib/desktop/assets.ts
import { invoke } from '@tauri-apps/api/core';
import type { TauriActionResult } from '@/lib/desktop/admin-types';
import { isTauriBackendAvailable } from '@/lib/desktop/runtime';

function requireTauri(): void {
  if (!isTauriBackendAvailable()) {
    throw new Error('Tauri assets API requires the desktop shell');
  }
}

export async function tauriAssetsCreateAsset(data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('assets_create_asset_cmd', { data });
}

export async function tauriAssetsGetAsset(assetId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown> | null>>('assets_get_asset_cmd', { assetId });
}

export async function tauriAssetsGetAssetsByOrg(orgId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('assets_get_assets_by_org_cmd', { orgId });
}

export async function tauriAssetsGetAssetsByStatus(orgId: string, status: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('assets_get_assets_by_status_cmd', {
    orgId,
    status,
  });
}

export async function tauriAssetsGetAssetsByUser(userId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('assets_get_assets_by_user_cmd', {
    userId,
  });
}

export async function tauriAssetsGetAssetByTag(tag: string, orgId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown> | null>>('assets_get_asset_by_tag_cmd', {
    tag,
    orgId,
  });
}

export async function tauriAssetsUpdateAsset(assetId: string, data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown> | null>>('assets_update_asset_cmd', {
    assetId,
    data,
  });
}

export async function tauriAssetsAssignAssetToUser(assetId: string, userId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown> | null>>('assets_assign_asset_to_user_cmd', {
    assetId,
    userId,
  });
}

export async function tauriAssetsDeleteAsset(assetId: string) {
  requireTauri();
  return invoke<TauriActionResult<boolean>>('assets_delete_asset_cmd', { assetId });
}

File: src-tauri/src/commands/assets.rs
use serde_json::Value;
use tauri::State;

use crate::admin::service::ActionResult;
use crate::assets::service::{
    assign_asset_to_user, create_asset, delete_asset, get_asset, get_asset_by_tag, get_assets_by_org,
    get_assets_by_status, get_assets_by_user, get_assets_for_current_org, update_asset,
};
use crate::auth::session::SessionState;
use crate::state::AppState;

#[tauri::command]
pub fn assets_create_asset_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    data: Value,
) -> ActionResult<Value> {
    create_asset(&app, &session, data)
}

#[tauri::command]
pub fn assets_get_asset_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    asset_id: String,
) -> ActionResult<Value> {
    get_asset(&app, &session, &asset_id)
}

#[tauri::command]
pub fn assets_get_assets_by_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
) -> ActionResult<Vec<Value>> {
    get_assets_by_org(&app, &session, &org_id)
}

#[tauri::command]
pub fn assets_get_assets_for_current_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    limit: Option<usize>,
) -> ActionResult<Vec<Value>> {
    get_assets_for_current_org(&app, &session, limit.unwrap_or(500))
}

#[tauri::command]
pub fn assets_get_assets_by_status_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
    status: String,
) -> ActionResult<Vec<Value>> {
    get_assets_by_status(&app, &session, &org_id, &status)
}

#[tauri::command]
pub fn assets_get_assets_by_user_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    user_id: String,
) -> ActionResult<Vec<Value>> {
    get_assets_by_user(&app, &session, &user_id)
}

#[tauri::command]
pub fn assets_get_asset_by_tag_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    tag: String,
    org_id: String,
) -> ActionResult<Value> {
    get_asset_by_tag(&app, &session, &tag, &org_id)
}

#[tauri::command]
pub fn assets_update_asset_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    asset_id: String,
    data: Value,
) -> ActionResult<Value> {
    update_asset(&app, &session, &asset_id, data)
}

#[tauri::command]
pub fn assets_assign_asset_to_user_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    asset_id: String,
    user_id: String,
) -> ActionResult<Value> {
    assign_asset_to_user(&app, &session, &asset_id, &user_id)
}

#[tauri::command]
pub fn assets_delete_asset_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    asset_id: String,
) -> ActionResult<bool> {
    delete_asset(&app, &session, &asset_id)
}
